src/cli/index.js

```diff
@@ -382,7 +343,7 @@ config.init({
   httpsProxy: commander.httpsProxy,
   networkConcurrency: commander.networkConcurrency,
   nonInteractive: commander.nonInteractive,
-  commandName,
+  commandName: commandName === 'run' ? commander.args[0] : commandName,
```

Comment:
Not sure we can do this; `yarn version` and `yarn run version` would both end up with a `version` commandName, which is semantically wrong because they have different behaviors. If we want to support user-defined commands here, then it should be in a separate field.